{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 41,
      "id": "bfdc4128-7b87-4312-9985-43ba1697a9ba",
      "metadata": {
        "id": "bfdc4128-7b87-4312-9985-43ba1697a9ba"
      },
      "outputs": [],
      "source": [
        "from random import randint\n",
        "from copy import deepcopy\n",
        "\n",
        "infile = open('input1.txt', 'r')\n",
        "\n",
        "#Create adjacency list\n",
        "def CreateAdjacencyList(infile):\n",
        "    \n",
        "    adj_list = {}\n",
        "    circuit_max = 0\n",
        "    for line in infile:\n",
        "        node = line.strip('\\n')\n",
        "        node = node.replace(' -> ', ' ')\n",
        "        node = node.split(' ')\n",
        "        adj_list.setdefault(node[0], [])\n",
        "        for number in node[1].split(','):\n",
        "            adj_list[node[0]].append(number)\n",
        "            circuit_max += 1\n",
        "    \n",
        "    return adj_list, circuit_max\n",
        "\n",
        "#Find start/end nodes\n",
        "def FindStart(red_adj_list):\n",
        "    \n",
        "    start = {}\n",
        "    for one in red_adj_list:\n",
        "        start.setdefault(one, 0)\n",
        "        start[one] += len(red_adj_list[one])\n",
        "\n",
        "    end = {}\n",
        "    for one in red_adj_list:\n",
        "        for two in red_adj_list[one]:\n",
        "            end.setdefault(two, 0)\n",
        "            end[two] += 1\n",
        "\n",
        "    for one in end:\n",
        "        try: \n",
        "            if start[one] != end[one]:\n",
        "                if start[one] > end[one]:\n",
        "                    start_node = one\n",
        "                if start[one] < end[one]:\n",
        "                    end_node = one\n",
        "        except KeyError:\n",
        "            end_node = one\n",
        "\n",
        "    for one in start:\n",
        "        try:\n",
        "            if end[one] != start[one]:\n",
        "                if end[one] < start[one]:\n",
        "                    start_node = one\n",
        "                if end[one] > start[one]:\n",
        "                    end_node = one\n",
        "        except KeyError:\n",
        "            start_node = one\n",
        "\n",
        "    red_adj_list[end_node] = []\n",
        "    \n",
        "    return red_adj_list, start_node\n",
        "\n",
        "def FindEulerianPath(infile):\n",
        "    \n",
        "    #Create adjacency list\n",
        "    adj_list, circuit_max = CreateAdjacencyList(infile)\n",
        "    \n",
        "    #Reduced adjacency list to keep track of traveled edges  \n",
        "    red_adj_list = {}            \n",
        "    red_adj_list = deepcopy(adj_list)\n",
        "    \n",
        "    #Find start node (graph must be directed/ubalanced)\n",
        "    red_adj_list, start_node= FindStart(red_adj_list)\n",
        "    \n",
        "    start = start_node\n",
        "    curr_vrtx = start_node\n",
        "    path = [curr_vrtx]\n",
        "    stack = []\n",
        "    circuit = []\n",
        "    while len(circuit) != circuit_max:\n",
        "    \n",
        "        if red_adj_list[curr_vrtx] != []: #If neighbors exist\n",
        "            stack.append(curr_vrtx)\n",
        "            pick = randint(0,len(red_adj_list[curr_vrtx])-1)\n",
        "            temp = deepcopy(curr_vrtx)\n",
        "            curr_vrtx = red_adj_list[temp][pick]\n",
        "            red_adj_list[temp].remove(curr_vrtx)\n",
        "        \n",
        "        else:\n",
        "            circuit.append(curr_vrtx)\n",
        "            curr_vrtx = stack[len(stack)-1]\n",
        "            stack.pop()\n",
        "    \n",
        "    #Formatting\n",
        "    path = start + '->'\n",
        "    for vrtx in circuit[::-1]:\n",
        "        path += (vrtx + '->')\n",
        "    return path.strip('->')\n",
        "with open(\"output.txt\", \"w\") as fd:\n",
        "    fd.write( FindEulerianPath(infile))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "id": "2b8db25f",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "11100100011000011001111100111101000000011000101111111110100110100001101111011001110010011001100100100111111000111100101001001011100101111010100110001110111100010010100001000111001100000111000001010111010010000011010110010110101011000000000101101111101101001011000110010101001010110110101111100001011101101100100001010001001101110101110000111101111110111011100111011000100010101011110011010100000100001110101010100011111010110100010110011011000010011100011010011101000110110111000101001111000000100100010000001111\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "'11100100011000011001111100111101000000011000101111111110100110100001101111011001110010011001100100100111111000111100101001001011100101111010100110001110111100010010100001000111001100000111000001010111010010000011010110010110101011000000000101101111101101001011000110010101001010110110101111100001011101101100100001010001001101110101110000111101111110111011100111011000100010101011110011010100000100001110101010100011111010110100010110011011000010011100011010011101000110110111000101001111000000100100010000001111'"
            ]
          },
          "execution_count": 44,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import sys\n",
        "import timeit\n",
        "import random\n",
        "from itertools import combinations,product,chain\n",
        "from collections import Counter,defaultdict\n",
        "\n",
        "def read_file(input_file):\n",
        "    '''\n",
        "    >>> k = read_file('test.k-universal_circular_string.4.txt')\n",
        "    '''\n",
        "    f = open(input_file)\n",
        "    data = f.read().strip()\n",
        "    f.close()\n",
        "    return int(data)\n",
        "\n",
        "def generate_binary(k):\n",
        "    return [bin(i)[2:].zfill(k) for i in range(2**k)]\n",
        "\n",
        "def form_relation(k):\n",
        "    nodes = generate_binary(k-1)\n",
        "    d = {}\n",
        "    for item in nodes:\n",
        "        add1 = item[1:]+'0'\n",
        "        add2 = item[1:]+'1'\n",
        "        d[item] = [add1,add2]\n",
        "    return d\n",
        "\n",
        "def form_cycle(data):\n",
        "    rgk = random.choice(list(data.keys()))\n",
        "    rgv = random.choice(data[rgk])\n",
        "    cycle = [rgk]\n",
        "    while len(data) != 0:\n",
        "        try:\n",
        "            cycle.append(rgv)\n",
        "            if len(data[rgk]) > 1:\n",
        "                data[rgk].remove(rgv)\n",
        "            else:\n",
        "                del data[rgk]\n",
        "            rgk = rgv\n",
        "            rgv = random.choice(data[rgk])\n",
        "            if rgv == cycle[0] and rgv == cycle[-1]:\n",
        "                break\n",
        "        except:\n",
        "            break\n",
        "    return (cycle,data)\n",
        "\n",
        "def form_unCycle(data,rgk):\n",
        "    choose = data[rgk]\n",
        "    rgv = random.choice(choose)\n",
        "    cycle = [rgk]\n",
        "    while len(data) != 0:\n",
        "        try:\n",
        "            cycle.append(rgv)\n",
        "            if len(data[rgk]) > 1:\n",
        "                data[rgk].remove(rgv)\n",
        "            else:\n",
        "                del data[rgk]\n",
        "            rgk = rgv\n",
        "            rgv = random.choice(data[rgk])\n",
        "            if rgv == cycle[0] and rgv == cycle[-1]:\n",
        "                break\n",
        "        except:\n",
        "            break\n",
        "    return (cycle,data)\n",
        "\n",
        "def fuse(Cycle,new_Cycle):\n",
        "    fuse_index = Cycle.index(new_Cycle[0])\n",
        "    return Cycle[:fuse_index]+new_Cycle+Cycle[fuse_index+1:]\n",
        "\n",
        "def eulerian_cycle(data,k):\n",
        "    Cycle,unCycle = form_cycle(data)\n",
        "    while len(unCycle) != 0:\n",
        "        keys = unCycle.keys()\n",
        "        potential = list(set(Cycle)&set(keys))\n",
        "        newStart = random.choice(potential)\n",
        "        new_Cycle,unCycle = form_unCycle(unCycle,newStart)\n",
        "        Cycle = fuse(Cycle,new_Cycle)\n",
        "    return Cycle\n",
        "\n",
        "def form_string(path):\n",
        "    string = ''.join([item[-1] for item in path[1:]])\n",
        "    return string\n",
        "\n",
        "def result():\n",
        "    k = 9\n",
        "    data = form_relation(k)\n",
        "    path = eulerian_cycle(data,k)\n",
        "    results = form_string(path)\n",
        "    print(results)\n",
        "    return results\n",
        "result()\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "6lb_vsWDwELC",
      "metadata": {
        "id": "6lb_vsWDwELC"
      },
      "source": [
        "Notes: We want to incorporate sequencing errors in the fragments f1 and f2 and get the confidence whether f1 and f2 belong from the same haplotype. Use some probablity to guess whether f1 and f2 belong to the same haplotype. \n",
        "\n",
        "Now once we have a probablity that f1 and f2 belong to the same haplotype, we want to check whether a variant is a false variant.\n",
        "\n",
        "What is a false variant ? \n",
        "False variant is a location where the alleles are homozygous but the sequencing \n",
        "assumes / wrongly interprets the site as heterozygous. \n",
        "Hitherto, in the fragments (for these false variants) there would be a substantial amount of fragments where this site has the same allele say 0 for \n",
        "both fragments coming from both H1 and H2 (as it is not really a variant hence same allele is observed irrespective sample from H1 or H2 (0, 0) in the given position).\n",
        "\n",
        "However for sequencing errors the errors are randomly distributed among different variants for different fragments.\n",
        "\n",
        "Problem statement, we want to identify these false variant locations for the \n",
        "sequenced data, as these become difficult to handle while performing Hap assembly.\n",
        "\n",
        "When we have a good coverage for a particular false variant, \n",
        "if we have fragment f1 which belongs to H1 and a fragment f2 which belongs to H2, the value for the variants are complement of each other except for the false variant. \n",
        "\n",
        "Now we dont know whether the fragment belongs to either H1 or H2, "
      ]
    }
  ],
  "metadata": {
    "colab": {
      "name": "HAP_simulation.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.6"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
